File: rotate_array.cpp
/* 
 * PS: Rotate an array by d elements clockwise
 * We were trying to solve in O(n), Auxilary space O(1)
 */

#include <iostream>
#include <numeric>
using namespace std;

void leftRotate(int* arr, int d, int n) {
    int g_c_d = gcd(n, d);
    for (int i=0; i<g_c_d; i++) {
        int j = i;
        int temp = arr[j];

        while (true) {
            int k = j + d;
            if (k >= n) {
                k -= n; // or k = k % n
            }

            if (k == i) {
                break;
            }
            arr[j] = arr[k];
            j = k;
        }
        arr[j] = temp;
    }
}

int main() {
    int arr[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};
    int n = sizeof(arr)/sizeof(arr[0]);

    leftRotate(arr, 4, n);
    for (int i=0; i<n; i++) {
        cout << arr[i] << " "; 
    }
    cout << endl;

    return 0;
}

File: RotateArray.java
import java.util.Arrays;

public class RotateArray {

    public static void main(String[] args) {
        int nums[] = { 1, 2, 3, 4, 5, 6, 7, 8 };
        int k = 3;

        rotate(nums, k);
        System.out.println(Arrays.toString(nums));
    }

    public static int gcd(int a, int b) {
        if (b == 0)
            return a;
        return gcd(b, a % b);
    }

    public static void rotate(int[] nums, int k) {
        int gcd = gcd(nums.length, k);
        for (int i = 0; i < gcd; i++) {
            int temp = nums[i];
            int j = i;
            while (true) {
                int next = (j + k) % nums.length;
                if (next == i)
                    break;
                nums[j] = nums[next];
                j = next;
            }
            nums[j] = temp;
        }
    }
}
